lib/Installation/ProductSelection/ProductSelectionController.pm

```diff
@@ -0,0 +1,56 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+
+# Summary: The class introduces business actions for Product Selection
+# Maintainer: QE YaST <qa-sle-yast@suse.de>
+
+package Installation::ProductSelection::ProductSelectionController;
+use strict;
+use warnings;
+use YuiRestClient;
+use Installation::ProductSelection::ProductSelectionPage;
+
+sub new {
+    my ($class, $args) = @_;
+    my $self = bless {}, $class;
+    return $self->init();
+}
+
+sub init {
+    my ($self) = @_;
+    $self->{ProductSelectionPage} = Installation::ProductSelection::ProductSelectionPage->new({
+            app                        => YuiRestClient::get_app(),
+            SLES_filter                => {label => 'SUSE Linux Enterprise Server 15 SP3'},
+            HPC_filter                 => {label => 'SUSE Linux Enterprise High Performance Computing 15 SP3'},
+            SLES_for_SAP_filter_filter => {label => 'SUSE Linux Enterprise Server for SAP Applications 15 SP3'},
+            SLED_filter                => {label => 'SUSE Linux Enterprise Desktop 15 SP3'},
+            SMGR_Server_filter         => {label => 'SUSE Manager Server 4.2'},
+            SMGR_Proxy_filter          => {label => 'SUSE Manager Proxy 4.2'},
+            SMGR_Retail_filter         => {label => 'SUSE Manager Retail Branch Server 4.2'}});
```

Comment:
- I guess these label filters can and should be moved to ProductSelectionPage, as only Page layer should has this knowledge.

    ```perl
    $self->{rb_SLED} = $self->{app}->radiobutton({label => 'SUSE Linux Enterprise Desktop 15 SP3});
    ```

- More severe issue that I see is hard-coded versions of the products. I know, that all the ids and labels on this page contain versions, but hard-coded versions that may be changed at any time will make maintaining more complex for us in future. I guess, we need to introduce something similar to the function, that we have in ruby client: https://github.com/qe-yast/ruby-yui-rest-client/blob/cc2951d99ec8db2cb581f5d42ff66ae6c9441d10/lib/yui_rest_client/widgets/base.rb#L93

   This function allows to collect all the elements with the same filter (e.g. YRadioButton) and then manipulate with them as needed using Perl (e.g. find the required item from the list of returned items using regex and then send the appropriate action).

Please, let me know if you need some details regarding this or want to discuss it.

Reply:
>     * I guess these label filters can and should be moved to ProductSelectionPage, as only Page layer should has this knowledge.
>       ```perl
>       $self->{rb_SLED} = $self->{app}->radiobutton({label => 'SUSE Linux Enterprise Desktop 15 SP3});
>       ```
> 
>     * More severe issue that I see is hard-coded versions of the products. I know, that all the ids and labels on this page contain versions, but hard-coded versions that may be changed at any time will make maintaining more complex for us in future. I guess, we need to introduce something similar to the function, that we have in ruby client: https://github.com/qe-yast/ruby-yui-rest-client/blob/cc2951d99ec8db2cb581f5d42ff66ae6c9441d10/lib/yui_rest_client/widgets/base.rb#L93
>       This function allows to collect all the elements with the same filter (e.g. YRadioButton) and then manipulate with them as needed using Perl (e.g. find the required item from the list of returned items using regex and then send the appropriate action).
> 
> 
> Please, let me know if you need some details regarding this or want to discuss it.

This is thought because the page probably would not change, just the way to search for the ids, so overwriting the controller in SP4 would just work. If we move to the Page then we need to override both controller and page when the page really don't change at all (functionally, same layout other strings). Make sense to you?

Reply:
This approach is thought because the page probably would not change, just the way to search for the ids, so overwriting the controller in SP4 would just work. If we move to the Page then we need to override both controller and page when the page really don't change at all (functionally, same layout other strings). Make sense to you?